src/codegen/compile.ml

```diff
@@ -5242,6 +5242,22 @@ module RTS_Exports = struct
       edesc = nr (FuncExport (nr rts_trap_fi))
     });
 
+    (* Keep a memory reserve when in update or init state. 
+       This reserve can be used by queries, composite queries, and upgrades. *)
+    let keep_memory_reserve_fi = E.add_fun env "keep_memory_reserve" (
+      Func.of_body env [] [I32Type] (fun env ->
+        Lifecycle.get env ^^
+        compile_eq_const Lifecycle.(int_of_state InUpdate) ^^
```

Comment:
Does this also cover heartbeat and timers?

Reply:
Yes, as far as I know. I also checked it with some local tests: Heartbeat and timers functions (defined by the user program) are executed as `InUpdate` (inferred from `Shared Write`).
Maybe @ggreif could confirm?

Reply:
I think that's correct. The only slight wrinkly is that composite_queries callbacks also execute in this mode, I think, so the callbacks will try to maintain the reserve. Does it matter?

Reply:
Good point, thanks for noticing: Are the callbacks of composite queries executed under `InUpdate` and later rolled back? I wonder whether there would be an easy way to also detect them and exempt them from the reserve.

Reply:
Just double checked: Composite queries are already correctly handled in the current version. They have state `InComposite` and can exhaust the memory reserve.

Reply:
Update: I now also detect callbacks of composite queries and allow them to also use the memory reserve.